### ISSUE TITLE
botbuilder-toybox-memories: a different take on state managment

### ISSUE THREAD

**Issue by Stevenic** [Microsoft/botbuilder-js#185]
I wanted to share out an alternate spin on state management that I cooked up over the weekend. I created a new package called [botbuilder-toybox-memories](https://github.com/Stevenic/botbuilder-toybox/tree/master/packages/botbuilder-toybox-middleware) which you can install from npm using `npm install botbuilder-toybox-memories@preview`.  There's also a [echobot sample](https://github.com/Stevenic/botbuilder-toybox/tree/master/samples/memories/echobot-ts) but I'll try to walk through the basic concepts below. I'd love feedback or to hear other thoughts or ideas this provokes.

I called the package **memories** as its the start of an effort to model bot memory (state management) in a more intelligent and realistic way.  In many ways it's similar to some of the other state ideas being floated around but some of the key things it adds is the ability to quickly learn what pieces of state the bot will need for a given turn (and then pre-load it in parallel) and the ability to automatically forget individual pieces of state after a period of time.  To illustrate the concepts I'll just jump into sample code that modifies the echobot to forget it's count after a period of time.

### Memory Scopes
The first thing you need to do is define your bots "Memory Scopes" and then add them to your bots middleware stack using a `MemoryScopeManager` class.  I'm adding just a `ConversationScope` below but there's other pre-defined scopes like `UserScope` and you can define any scope you want using the `MemoryScope` class.  Storage wise this is all just using M3's storage providers.

```JS
import { ConversationScope, MemoryScopeManager, MemoryScopeAccessor } from 'botbuilder-toybox-memories';

// Define memory scopes add memory manager middleware
const conversation = new ConversationScope(new MemoryStorage());
adapter.use(new MemoryScopeManager(conversation));
```

You can give the `MemoryScopeManager` as many scopes to manage as you want without having to worry about performance optimizations around reading in particular scopes in parallel.  All reads are deferred until the state is needed by default and the manager watches the activities flowing through your bot and then quickly learns which scopes you use for a given activity type and then it will begin pre-loading the appropriate scope for you within a few activities. So if you never touch state when doing a proactive message it will figure that out and won't pre-load any state when you send a proactive message.

### Memory Fragments
A completely new concept is the notion of a "Memory Fragment". Scopes start off completely empty and you have to define the fragments of memory they contain.  While this may feel a little heavy handed its actually quite powerful because it lets you do things like define default values and how long things should live.

```JS
// Define memory fragments
conversation.fragment('count', 0).forgetAfter(10);
```

In the fragment above I'm defining a "count" property that starts with a value of 0 and should automatically be forgotten after 10 seconds.  To consume the fragment we can update our echobots code to look like this:

```JS
    adapter.processActivity(req, res, async (context: MyContext) => {
        if (context.activity.type === 'message') {
            let count = await context.conversation.get('count') as number;
            await context.sendActivity(`${++count}: You said "${context.activity.text}"`);
            await context.conversation.set('count', count);
        } else {
            await context.sendActivity(`[${context.activity.type} event detected]`);
        }
    });
```

The `MemoryScopeManager` automatically extends the context object with a `MemoryScopeAccessor` property for each scope its managing (you can change the property name.) The accessor contains async methods for getting, setting, and re-setting a fragment.  Since I'm using a primitive `number` I actually need to `set()` the updated value.  If my fragment was instead an `object` I could just update the fragments properties without needing to call `set()`.  

It's perfectly reasonable to have single fragment that represents your bots entire conversation state. The primary advantage of supporting multiple fragments within a scope is that it lets you a) have different life times for things, and b) it lets components partition their state so that they're not clobbering your bots state and vice-versa.  In fact the `MemoryFragment` that gets defined can be used as an accessor in its own right so to give a component access to a block of memory you can do this:

```JS
// Define scopes
const user = new UserScope(storage);
const conversation = new ConversationScope(storage);
adapter.use(new MemoryScopeManager(user, conversation));

// Define fragments
const userProfile = user.fragment('userProfile');
const firstRunState = conversation.fragment('firstRunState');

// Add first run middleware
adapter.use(new FirstRunMiddleware(firstRunState, userProfile, config));
```

That block of code is actually what largely inspired this idea and adds a `FirstRunMiddleware` giving it a block of memory to persist its state too and the ability to read-write the users profile. The middleware is isolated from touching any of the bots other memory.  Those fragments can even be strongly typed if you're using TypeScript.
 
### TypeScript Users
If you're using TypeScript you'll want to extend the `TurnContext` interface so that you'll intellisense:

```JS
// Extend TurnContext interface
interface MyContext extends TurnContext {
    conversation: MemoryScopeAccessor;
}
```

**Comment by cleemullins** [comment]
This looks like a great starting point for a conversation to have post-build.

**Comment by Stevenic** [comment]
Closing as we have a new state management system that builds on this.
